src/lem/circuit.rs

```diff
@@ -1045,39 +1212,58 @@ impl Func {
-                                acc.and_then(|acc| b.get_value().map(|b| acc && !b))
-                            });
-                            let has_match = Boolean::Is(AllocatedBit::alloc(
-                                &mut cs.namespace(|| "_.allocated_bit"),
-                                default,
+                    if let Some(def) = def {
+                        let default = selector.iter().fold(not_dummy.get_value(), |acc, b| {
+                            acc.and_then(|acc| b.get_value().map(|b| acc && !b))
+                        });
+                        let has_match = Boolean::Is(AllocatedBit::alloc(
+                            &mut cs.namespace(|| "_.allocated_bit"),
+                            default,
+                        )?);
+                        for (i, (lit, _)) in cases.iter().enumerate() {
+                            let lit_ptr = lit.to_ptr_cached(g.store);
+                            let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
+                            let lit_tag = g.store.hash_ptr(&lit_ptr)?.tag.to_field::<F>();
+                            // TODO optimize and make it more understandable
+                            let choose_unequal = Boolean::Is(AllocatedBit::alloc(
+                                &mut cs.namespace(|| format!("{i}.select_unequal")),
+                                match_lit.tag().get_value().map(|tag| tag != lit_tag),
                             )?);
-                            for (i, (lit, _)) in cases.iter().enumerate() {
-                                let lit_ptr = lit.to_ptr(g.store);
-                                let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
-                                implies_unequal_const(
-                                    &mut cs.namespace(|| format!("{i} implies_unequal")),
-                                    &has_match,
-                                    &match_lit,
-                                    lit_hash,
-                                )?;
-                            }
-
-                            selector.push(has_match.clone());
-
-                            recurse(
-                                &mut cs.namespace(|| "_"),
-                                def,
+                            let choose_unequal_def = and(
+                                &mut cs.namespace(|| format!("{i} implies_unequal and")),
+                                &choose_unequal,
                                 &has_match,
-                                next_slot,
-                                bound_allocations,
-                                preallocated_outputs,
-                                g,
+                            )?;
+                            let choose_unequal_def_not = and(
+                                &mut cs.namespace(|| format!("{i} implies_unequal and not")),
+                                &choose_unequal.not(),
```

Comment:
Is this `choose_equal_def` then? If so, it all might be clearer if you defined `choose_equal` as the primary value, then derived `choose_unequal_def` and `choose_equal_def` from it. As written, this is quite hard to parse.

Reply:
Also, I think some of these namespaces need to be updated.

Reply:
`Match` was replaced by the simpler `MatchSymbol`